e2core/backend/satbackend/watcher.go

```diff
@@ -86,7 +117,10 @@ func (w *watcher) terminate() {
 		instance.cxl(errTerminateAll)
 
 		delete(w.instances, p)
+		w.instancesRunning.Add(-1)
 	}
+
+	w.instancesRunning.Wait()
```

Comment:
It's hard to tell from the diff... why are we waiting on this?

Reply:
when a new process is started with `exec.Start()`, one of the things that return from that function is a wait function.

That wait function is then placed in a goroutine which blocks on the specific command's (process's) wait, and only decrements the waitgroup once it gets past it.

This wait waits until all of the instances are terminated, which means all of the wait functions unblock, the wait group reaches zero.

At the end of the day it cleans up all the goroutines, so we won't have orphan goroutines that aren't attached to anything any more because the world was pulled out from under them. Or in other words, their parent process existed / finished.

Reply:
Heh I meant 'in what scenario would we want to wait on this' but I see now it's when we want to terminate the entire watcher. That makes sense.